frontend/public/module/ws-factory.ts

```diff
@@ -0,0 +1,344 @@
+      this.flushCanceler = setInterval(
+        this.flushMessageBuffer.bind(this),
+        this.options.bufferFlushInterval || 500,
+      );
+    }
+  }
+
+  private reconnect() {
+    if (this.connectionAttempt || this.state === 'destroyed') {
+      return;
+    }
+
+    let delay = 1000;
+
+    const attempt = () => {
+      if (!this.options.reconnect || this.state === 'open') {
+        clearTimeout(this.connectionAttempt);
+        this.connectionAttempt = null;
+        return;
+      }
+      if (this.options.timeout && delay > this.options.timeout) {
+        clearTimeout(this.connectionAttempt);
+        this.connectionAttempt = null;
+        this.destroy();
+        return;
+      }
+
+      this.connect();
+      delay = Math.round(Math.min(1.5 * delay, 60000));
+      this.connectionAttempt = setTimeout(attempt, delay);
+      // eslint-disable-next-line no-console
+      console.log(`attempting reconnect in ${delay / 1000} seconds...`);
+    };
+
+    this.connectionAttempt = setTimeout(attempt, delay);
+  }
+
+  private connect() {
+    const that = this;
+    this.state = 'init';
+    this.messageBuffer = [];
+    try {
+      this.ws = new WebSocket(this.url, this.options.subprotocols);
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.error('Error creating websocket:', e);
+      this.reconnect();
+      return;
+    }
+
+    this.ws.onopen = function() {
+      // eslint-disable-next-line no-console
+      console.log(`websocket open: ${that.id}`);
+      that.state = 'open';
+      that.triggerEvent('open', undefined);
+      if (that.connectionAttempt) {
+        clearTimeout(that.connectionAttempt);
+        that.connectionAttempt = null;
+      }
+    };
+    this.ws.onclose = function(evt: CloseEvent) {
+      // eslint-disable-next-line no-console
+      console.log(`websocket closed: ${that.id}`, evt);
+      that.state = 'closed';
+      that.triggerEvent('close', evt);
+      that.reconnect();
+    };
+    this.ws.onerror = function(evt: Event) {
+      // eslint-disable-next-line no-console
+      console.log(`websocket error: ${that.id}`);
+      that.state = 'error';
+      that.triggerEvent('error', evt);
+    };
+    this.ws.onmessage = function(evt: Parameters<typeof WebSocket.prototype.onmessage>[0]) {
+      const msg = that.options?.jsonParse ? JSON.parse(evt.data) : evt.data;
+      // In some browsers, onmessage can fire after onclose/error. Don't update state to be incorrect.
+      if (that.state !== 'destroyed' && that.state !== 'closed') {
+        that.state = 'open';
+      }
+      that.triggerEvent('message', msg);
+    };
+  }
+
+  private registerHandler(type: string, fn: RegisterHandler) {
+    if (this.state === 'destroyed') {
+      return;
+    }
+    this.handlers[type].push(fn);
+  }
+
+  /**
+   * Invoke all registered handler callbacks for a given event type.
+   */
+  private invokeHandlers(type: string, data: MessageDataType) {
```

Comment:
```suggestion
  private invokeHandlers(type: keyof WSHandlers, data: MessageDataType) {
```
Sorry, I missed this. I think this helps keep internal types clean too.

(without this, `handlers` is `any[]` and `h` is of type `any`